02-peer-protocol.md

```diff
@@ -545,6 +545,42 @@ considered to be the channel funder (the sender of `open_channel`).
 The quiescence effect is exactly the same as if one had replied to the
 other.
 
+### `go_on`
+
+1. type: TODO (`go_on`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * **NOTE FOR REVIEWERS** do we need to include initiator here? I think not.
+
+#### Requirements
+
+The sender of `go_on`:
+  - MUST NOT send `go_on` for a channel that is NOT quiescent.
+  - MUST NOT send `go_on` if it is not the quiescence "initiator".
+  - MUST NOT send `go_on` twice without an interceding full quiescence
+    negotiation.
+  - MUST send `go_on` once any dependent protocol has completed.
+  - MAY send `update_*` messages afterwards.
```

Comment:
Shouldn't you wait to receive your peer's `go_on` (which acts as an ACK) before resuming updates? This would more clearly mark a collaborative exit of the `quiescence` state. If we don't cleanly exit that way, we should disconnect before resuming updates.

Reply:
I can tighten the requirement here, but I'm trying to figure out what problem it solves. This only really impacts the initiator so we can examine that.

If I'm the initiator, and I send `go_on` and do not wait before sending new `update_*` messages, the responder has two options. Due to the protocol assumption of in-order message processing, any subsequent updates must be processed after the `go_on`. As the responder, either I'm ready to process them, in which case I send my `go_on` before anything else as an ack and then proceed to handle the rest of the buffer, or I am not ready in which case I'll just hang up.

I think it's tempting to think that these messages are "commands" for the peer, but since the whole protocol is asynchronous we can't really think of things that way. I view `stfu` and `shutdown` as polite requests for your peer do do something but what they *really* are is declarations for what to expect from your future behavior. Under this mental model we don't really need to wait for the responder.

Reply:
It's simply cleaner from a state machine design point of view (which is why we're adding a `go_on` message in the first place). If you don't wait for your your peer to respond with `go_on` before sending updates, you're in a slightly weird state after sending the first `go_on`: you're not back to your normal, out-of-quiescence state because you need to handle receiving your peer's `go_on` message. So you'll need to import all of the message semantics from your "normal" state into that "waiting for peer's go_on" state plus a handler for receiving `go_on`, at which point you go back to your "normal" state and quiescence is completely exited.

Eclair uses a strict finite state machine design for most of its components, which shows this design weirdness. But if you feel strongly about it I don't really mind, because your current requirements also allow us to wait for the responding `go_on` before sending updates, which is what we'll implement.

Reply:
> So you'll need to import all of the message semantics from your "normal" state into that "waiting for peer's go_on" state plus a handler for receiving go_on, at which point you go back to your "normal" state and quiescence is completely exited.

Yeah I can see what you're saying where if you go for the strict FSM design and the only valid exit of quiescence is send/receive. I have preferred not to think about it as a strict FSM due to the asynchronous nature of the rest of the protocol.

While we have a total ordering of all messages on each half of the connection, we don't have a total ordering of all messages exchanged since there are many valid zips. As a result, as the responder, I have no way to guarantee you've received my `go_on` after I send it. I suppose if I don't ever send it I can guarantee you *haven't* received it. So maybe you're saying that your ACK of my `go_on` *response* is the resumption of `update_*` messages.

I actually don't feel strongly about it! And int the process of responding to this comment I'm finding myself convinced by this line of reasoning.

Reply:
> As a result, as the responder, I have no way to guarantee you've received my go_on after I send it.

Sure, but I have the guarantee that if you receive my future `update_*` messages, you must have received my `go_on` first so your FSM cleanly got back to the normal, non-quiescent channel state. I think that should be enough?

> So maybe you're saying that your ACK of my go_on response is the resumption of update_* messages.

Yes exactly, or the fact that if you resume sending `update_*` messages I won't disconnect on you.

> I actually don't feel strongly about it! And int the process of responding to this comment I'm finding myself convinced by this line of reasoning.

I don't feel too strongly about this either, as we still have the option of waiting for our peer's `go_on` before sending our `update_*` messages with the way you currently worded the spec. So it's your choice!

Reply:
I'll update as I do believe your suggestion makes it more likely that people will implement with predictable behavior.